daft/runners/ray_tracing.py

```diff
@@ -0,0 +1,666 @@
+            self.has_written_event = True
+
+    def write_event(self, event: dict[str, Any], ts: int | None = None) -> int:
+        """Write a single trace event to the file
+
+        Args:
+            event: The event data to write
+            ts: Optional timestamp override. If None, current time will be used
+
+        Returns:
+            The timestamp that was used for the event
+        """
+        if self.file is not None:
+            if ts is None:
+                ts = int((time.time() - self.start) * 1000 * 1000)
+            if self.has_written_event:
+                self.file.write(",\n")
+            json.dump(
+                {
+                    **event,
+                    "ts": ts,
+                },
+                self.file,
+            )
+            self.has_written_event = True
+        return ts or -1
+
+    def write_footer(self) -> None:
+        if self.file is not None:
+            # Close with a closing square bracket
+            self.file.write("]")
+
+
+class RunnerTracer:
+    def __init__(self, file: TextIO | None, metrics_actor: ray_metrics.MetricsActorHandle | None):
+        start = time.time()
+        self._start = start
+        self._stage_start_end: dict[int, tuple[int, int]] = {}
+        self._task_id_to_location: dict[str, tuple[int, int]] = {}
+        self._task_event_idx = 0
+
+        self._metrics_actor = metrics_actor
+
+        self._writer = TraceWriter(file, start)
+        self._writer.write_header()
+
+    def _write_event(self, event: dict[str, Any], ts: int | None = None) -> int:
+        try:
+            return self._writer.write_event(event, ts)
+        except (json.JSONDecodeError, TypeError) as e:
+            logger.exception("Failed to serialize event to JSON: %s", e)
+            return ts or -1
+        except OSError as e:
+            logger.exception("Failed to write trace event to file: %s", e)
+            return ts or -1
+
+    def _write_metadata(self, metadata_event: dict[str, Any]) -> None:
+        try:
+            return self._writer.write_metadata(metadata_event)
+        except (json.JSONDecodeError, TypeError) as e:
+            logger.exception("Failed to serialize metadata to JSON: %s", e)
+            return
+        except OSError as e:
+            logger.exception("Failed to write trace metadata to file: %s", e)
+            return
+
+    def _flush_task_metrics(self):
+        if self._metrics_actor is not None:
+            task_events, new_idx = self._metrics_actor.get_task_events(self._task_event_idx)
+            self._task_event_idx = new_idx
+
+            for task_event in task_events:
+                if isinstance(task_event, ray_metrics.StartTaskEvent):
+                    # Record which pid/tid to associate this task_id with
+                    self._task_id_to_location[task_event.task_id] = (task_event.node_idx, task_event.worker_idx)
+
+                    start_ts = int((task_event.start - self._start) * 1000 * 1000)
```

Comment:
I would factor this to a utility `def start_ts(self, task_event_start: float) -> int` since you use it in multiple places.

Reply:
Refactored 2 places where this was used